src/graph/Graph.java

```diff
@@ -0,0 +1,155 @@
+package graph;
+
+import java.util.*;
+
+public class Graph {
+
+  private Map<Integer, Vertex> vertices;
+  private String name;
+
+  public Graph() {
+    this("unnamed");
+  }
+
+  public Graph(String name) {
+    this.vertices = new HashMap<>();
+    this.name = name;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public void setName(String name) {
+    this.name = name;
+  }
+
+  public Vertex get_node(int id) {
+    return this.vertices.get(id);
+  }
+
+  public Set<Integer> node_ids() {
+    return vertices.keySet();
+  }
+
+  public Collection<Vertex> nodes() {
+    return vertices.values();
+  }
+
+  public Vertex add_node(Vertex v) {
+    this.vertices.put(v.getId(), v);
+    return v;
+  }
+
+  public Vertex add_node(int id) {
+    Vertex node = new Vertex(id);
+    return this.add_node(node);
+  }
+
+  public Vertex add_node() {
+    if (this.is_empty()) {
+      return this.add_node(0);
+    }
+    int max_id = Collections.max(this.vertices.keySet());
+    return this.add_node(max_id + 1);
+  }
+
+  public void remove_node(int id) {
+    for (Vertex v : this.nodes()) {
+      v.remove_neighbor_by_id(id);
+    }
+    this.vertices.remove(id);
+  }
+
+  public void remove_node(Vertex v) {
+    this.remove_node(v.getId());
+  }
+
+  public void add_edge(Vertex node_u, Vertex node_v) {
+    node_u.add_neighbor(node_v);
+    node_v.add_neighbor(node_u);
+  }
+
+  public void add_edge(int u, int v) {
+    Vertex node_u = this.vertices.get(u);
+    Vertex node_v = this.vertices.get(v);
+    if (node_u == null) {
+      node_u = this.add_node(u);
+    }
+    if (node_v == null) {
+      node_v = this.add_node(v);
+    }
+    this.add_edge(node_u, node_v);
+  }
+
+  public void add_edge(Edge e) {
```

Comment:
I would say: make each adjacency list in each node contain Edge object, instead of just neighboured Vertex, we would sometimes require Edge's attribute in the algorithm.